modules/ROOT/pages/database/database-connector-examples.adoc

```diff
@@ -0,0 +1,737 @@
+
+For example, when deleting, many rows could match the criteria and get deleted, but only one criteria `(POSITION = X)` is provided.
+The same concept applies for update; if you use `UPDATE PRODUCTS set PRICE = PRICE * 0.9 where PRICE > :price`, you may be applying a
+ 10% discount on many products, but the price input parameter only accepts one value.
+
+If you want to apply different discount rates on products that have different prices, you can do so by executing many operations.
+For example, assume you have a payload which is a list of object of the following structure { price : number, discountRate: number}.
+From that point, you can do the following:
+
+[source,xml,linenums]
+----
+<foreach>
+  <db:update config-ref="dbConfig">
+    <db:input-parameters>
+     #[
+      {
+        'discountRate' : payload.discountRate,
+        'price' : payload.price,
+      }
+    ]
+    </db:input-parameters>
+    <db:sql>
+      UPDATE PRODUCTS set PRICE = PRICE * :discountRate where PRICE > :price
+    </db:sql>
+  </db:update>
+</foreach>
+----
+
+The previous operation accomplishes the task at hand but is inefficient. For each element in the list, one query needs to be executed.
+As such, for each element, the operation must:
+
+* Parse the query
+* Resolve parameters
+* Acquire a connection to the database (either by getting one for the pool or establishing a new one)
+* Pay all the network overhead
+* The RBMS has to process the query and apply changes
+* Release the connection
+
+You can avoid all the above steps by doing a bulk operation. There is only one query in that example; the update statement is constant, not dynamic.
+The difference is that you supply a different set of parameters on each iteration.
+
+Bulk operations allow you to run a single query using a set of parameters values.
+Bulk operations are not a shortcut for the same `<foreach>` as noted in the previous example; they use features on the JDBC API so that:
+
+* The query is parsed only once
+* Only one database connection is required since a single statement is executed
+* Network overhead is minimized
+* RBDMS can execute the bulk operation atomically
+
+For these use cases, the connector offers three operations: `<db:bulk-insert>`, `<db:bulk-update>` and `<db:bulk-delete>`.
+
+These operations are similar to their single counterparts, except that instead of receiving input parameters as key-value pairs,
+the operations expect them as a list of key-value pairs.
+
+For example:
+
+[source,xml,linenums]
+----
+<db:bulk-insert config-ref="dbConfig" >
+  <db:bulk-input-parameters>
+    #[[{'id': 2, 'name': 'George', 'lastName': 'Costanza'}, {'id': 3, 'name': 'Cosmo', 'lastName': 'Kramer'}]]
+  </db:bulk-input-parameters>
+  <db:sql>
+    insert into customers (id, name, lastName) values (:id, :name, :lastName)
+  </db:sql>
+</db:bulk-insert>
+----
+
+= Example: Database Connector Data Types Reference - Mule 4
```

Comment:
Check your heading levels throughout.

Reply:
Can make h1 to h2 and h2 to h3? Will ping you.